### PR TITLE
Allow contents of Paragraph to be "connected" to a meta custom field

### PR DESCRIPTION
**Pull request by michalczaplinski**
Following up on work in #51375

This PR is a counterpart to **and** stacked on top of #53241 because we need some code from there to make it work. So #53241  should be merged first.

## What does this do?

We're adding a way to "connect" the contents of the Paragraph block to a meta custom field.

https://github.com/WordPress/gutenberg/assets/5417266/8f9ddba1-f4e5-4914-b236-422fc7f4c26b

### CODE REVIEW

File: lib/experimental/custom-sources/meta.php [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ * @package gutenberg
+ */
+
+return array(
+	'name'         => 'meta',
+	'apply_source' => function ( $block_content, $block_instance, $meta_field, $attribute_config ) {
+		// We should probably also check if the meta field exists but for now it's okay because
+		// if it doesn't, `get_post_meta()` will just return an empty string.
+		$meta_value = get_post_meta( $block_instance->context['postId'], $meta_field, true );
+		$p          = new WP_HTML_Tag_Processor( $block_content );
```
Comment (cbravobernal) [inline lib/experimental/custom-sources/meta.php]: `$p` and `$p2` are not quite descriptive 😅

File: lib/experimental/custom-sources/meta.php [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+		$tag_name = $p->get_tag();
+		$markup   = "<$tag_name>$meta_value</$tag_name>";
+		$p2       = new WP_HTML_Tag_Processor( $markup );
+		$p2->next_tag();
+		$names = $p->get_attribute_names_with_prefix( '' );
+		foreach ( $names as $name ) {
+			$p2->set_attribute( $name, $p->get_attribute( $name ) );
+		}
+
+		return $p2 . '';
```
Comment (cbravobernal) [inline lib/experimental/custom-sources/meta.php]: Do we need `get_updated_html()` or `apply_attributes_updates()` here?

File: lib/experimental/custom-sources/meta.php (hunk 1 above)
Comment (michalczaplinski) [reply lib/experimental/custom-sources/meta.php]: true 😅 I'll change that

File: lib/experimental/custom-sources/meta.php (hunk 2 above)
Comment (michalczaplinski) [reply lib/experimental/custom-sources/meta.php]: That's a good catch! I've just double checked and we don't need the `get_updated_html()` because `WP_HTML_Tag_Processor::__toString()` calls `get_updated_html()` automatically ([link](https://developer.wordpress.org/reference/classes/wp_html_tag_processor/__tostring/#source)).

I don't know if we need `apply_attributes_updates()` because I don't understand its use case. 

@dmsnell Could you enlighten me on what `apply_attributes_updates()` does? Is it always necessary to call it after calling `p->set_attribute()`? Or it only necessary in some circumstances?

Comment (dmsnell) [reply lib/experimental/custom-sources/meta.php]: userspace should never worry about calling `apply_attribute_updates()`. that should be an internal-use only function

File: lib/experimental/custom-sources/meta.php [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+		$tag_name = $tags->get_tag();
+		$markup   = "<$tag_name>$meta_value</$tag_name>";
+		$updated_tags       = new WP_HTML_Tag_Processor( $markup );
+		$updated_tags->next_tag();
+		$names = $tags->get_attribute_names_with_prefix( '' );
+		foreach ( $names as $name ) {
+			$updated_tags->set_attribute( $name, $tags->get_attribute( $name ) );
+		}
+
+		return $updated_tags->get_updated_html() . '';
```
Comment (dmsnell) [inline lib/experimental/custom-sources/meta.php]: curious what the ` . ''` is here for. were you running into errors that you're trying to cover?

File: packages/block-library/src/paragraph/block.json [hunk 4]
```diff
@@ -28,6 +29,14 @@
...
 			"type": "string",
 			"enum": [ "ltr", "rtl" ]
+		},
+		"connections": {
+			"type": "object",
+			"attributes": {
+				"content": {
+					"source": "meta_fields"
+				}
+			}
```
Comment (SantosGuillamot) [inline packages/block-library/src/paragraph/block.json]: Do we need to define the "connections" attribute? Ideally, I would like to explore the possibility of being able to populate that block attribute without having to define it in the `block.json`.  For example, that's what's happening with styles, right? Maybe it makes more sense to just have `supports.connections: true`?

File: lib/experimental/blocks.php [hunk 5]
```diff
@@ -121,3 +121,69 @@ function gutenberg_register_metadata_attribute( $args ) {
...
+	 */
+	function gutenberg_render_custom_sources( $block_content, $block, $block_instance ) {
+		$meta_custom_source = require __DIR__ . '/custom-sources/meta.php';
+		$block_type         = $block_instance->block_type;
+
+		// Whitelist of the block types that support custom sources
+		// Currently, we only allow the Paragraph and Image blocks to use custom sources.
+		if ( ! in_array( $block['blockName'], array( 'core/paragraph', 'core/image' ), true ) ) {
+			return $block_content;
+		}
```
Comment (SantosGuillamot) [inline lib/experimental/blocks.php]: Would it make sense to create a whitelist of block attributes instead of just blocks? Both in the paragraph and image, there are block attributes that won't work with the connections.

File: lib/experimental/blocks.php [hunk 6]
```diff
@@ -121,3 +121,69 @@ function gutenberg_register_metadata_attribute( $args ) {
...
+if ( $gutenberg_experiments && array_key_exists( 'gutenberg-connections', $gutenberg_experiments ) ) {
+	/**
+	 * Renders the block meta attributes.
+	 *
+	 * @param string   $block_content Block Content.
+	 * @param array    $block Block attributes.
+	 * @param WP_Block $block_instance The block instance.
+	 */
+	function gutenberg_render_custom_sources( $block_content, $block, $block_instance ) {
+		$meta_custom_source = require __DIR__ . '/custom-sources/meta.php';
```
Comment (SantosGuillamot) [inline lib/experimental/blocks.php]: I'm a bit confused about the "custom sources" naming. Is this what we are doing? Are we rendering custom sources? If I understood it correctly, we are:

* Processing the connections block attribute.
* Getting the value from the connections source. Meta in this case using `get_post_meta`.
* Replacing the HTML according to the block attributes source. `html` in the case of the paragraph.

In my opinion, this last step should be independent of the connections source. If I get the value from a different source, I would still have to replace the same HTML, but with a different value. Is this right?

Just to have more context, I believe in the future this distinction will be more relevant:

* **Connections sources**: From where we get the value. Right now it is only meta, but it could be drupal, parent-block, my-custom-database...
* **Block attributes source**: Dictates which part of the HTML needs to be replaced. Thanks to the block attributes source, it's possible to know the HTML we have to replace. If it is `html` we will replace the inner content. If it is `attribute` we will replace an HTML attribute.

That's why it makes sense to separate the code as well.

**Review by SantosGuillamot — COMMENTED** [review]
I left some comments to discuss. Apart from that, I believe there are some styles broken.

File: lib/experimental/custom-sources/meta.php (hunk 2 above)
Comment (michalczaplinski) [reply lib/experimental/custom-sources/meta.php]: gotcha. Thanks @dmsnell !

File: lib/experimental/custom-sources/meta.php (hunk 3 above)
Comment (michalczaplinski) [reply lib/experimental/custom-sources/meta.php]: Good catch. I hate to say this, but I'm not sure. Either it was something that Copilot suggested to me or I copied this from Riad's PR. Either way, I'll remove it because we don't need it, thanks!

File: lib/experimental/blocks.php (hunk 6 above)
Comment (michalczaplinski) [reply lib/experimental/blocks.php]: > * Processing the connections block attribute.
> * Getting the value from the connections source. Meta in this case using `get_post_meta`.
> * Replacing the HTML according to the block attributes source. `html` in the case of the paragraph.

You're right, that's the right abstraction of the 3 steps that we perform. As it stands right now, the 2nd & 3rd step are not separated. It makes sense that they should be.

Also, some of the naming is a direct copy from Riad's PR (https://github.com/WordPress/gutenberg/pull/51375) which is not 100% accurate anymore. 

I ll change it.

File: lib/experimental/blocks.php (hunk 5 above)
Comment (michalczaplinski) [reply lib/experimental/blocks.php]: This serves as a "whitelist" of the block attributes:

https://github.com/WordPress/gutenberg/blob/f0bcb0fb04f9412fb67303f12c101c073a7bc0c6/packages/block-editor/src/hooks/custom-fields.js#L67-L68

Do you think that's sufficient?

Comment (michalczaplinski) [reply lib/experimental/blocks.php]: I suppose we could only check the `content` attribute of the Paragraph and the `url` of the Image, instead of iterating over the attributes on this line: 

https://github.com/WordPress/gutenberg/blob/60fa0613981cbd1cc45010cefdae40d4a49be509/lib/experimental/blocks.php#L161

Comment (SantosGuillamot) [reply lib/experimental/blocks.php]: I don't know what the best mechanism is, and I guess we will have to iterate it, but right now if I add a connection manually to the block (without using the Block Editor), I could include any block attribute and the PHP logic will try to replace that, right?

Apart from that, keep in mind that in the future we will probably include support for other Image block attributes like `title`. I would say that the easier it is to add block attributes to those lists the better. As we will be adding support for them progressively.

Comment (michalczaplinski) [reply lib/experimental/blocks.php]: > I don't know what the best mechanism is, and I guess we will have to iterate it, but right now if I add a connection manually to the block (without using the Block Editor), I could include any block attribute and the PHP logic will try to replace that, right?

Yep, that's true. OK, makes sense that we create an explicit, more centralized whitelist in the `blocks.php` file.

File: packages/block-library/src/paragraph/block.json (hunk 4 above)
Comment (michalczaplinski) [reply packages/block-library/src/paragraph/block.json]: Yeah, that makes sense Mario, let me make that change.

Comment (michalczaplinski) [reply packages/block-library/src/paragraph/block.json]: done in 367ee0c34a482c1199031f3768746c5fdcd12304

File: lib/experimental/blocks.php (hunk 5 above)
Comment (michalczaplinski) [reply lib/experimental/blocks.php]: done in 49d5cb6e3c6a167825af85cf00e40370cf767988

File: lib/experimental/blocks.php (hunk 6 above)
Comment (michalczaplinski) [reply lib/experimental/blocks.php]: The code should be better abstracted now to reflect those 3 separate steps. 1 & 3 are both inside of the same function but we can abstract it further when we add support for the Image block & other attributes. Then we can abstract the part that replaces the different attribute sources (`html`, `attribute`, etc.).

We should discuss consistent naming though of Step 2. Should it be:
- custom sources
- custom connections
- connection sources
- custom connection sources
- ...something else?

Comment (SantosGuillamot) [reply lib/experimental/blocks.php]: Yes, I believe the whole naming should be discussed. Another option I was thinking of was using "Dynamic data" instead of "Connections". So it could be "Dynamic data sources". But I didn't think enough about it. What doesn't quite fit me is the use of "custom". What does "custom" mean in this context? Which sources are not custom?

Do we need it for this initial experiment? Is it too difficult to change it afterwards?

Comment (michalczaplinski) [reply lib/experimental/blocks.php]: > Is it too difficult to change it afterwards?
Nope, it won't be difficult at all, I just wanted to start this conversation and have consistent naming even if it gets changed later 🙂 .

I guess "custom" means something like "different than whatever WP uses to get the content normally" but I agree that it's a bit meaningless.

I like "Connections" a little bit better than "Dynamic Data" because the latter can mean many things in many different contexts. Most data is dynamic rather than static in programming 🤓. I think "Connection Sources" sounds best so far. "Connections" alone is also a bit vague. "Connection Sources" implies "_data_ sources" which is roughly what we are referring to here.

Comment (michalczaplinski) [reply lib/experimental/blocks.php]: Maybe "block connections" is also a good name? We're **connecting stuff to blocks** after all.

File: packages/block-editor/src/hooks/custom-fields.js [hunk 7]
```diff
@@ -14,18 +14,18 @@ import { InspectorControls } from '../components';
...
  * @return {Object} Filtered block settings.
  */
 function addAttribute( settings ) {
 	if ( hasBlockSupport( settings, '__experimentalConnections', true ) ) {
-		// Gracefully handle if settings.attributes is undefined.
+		// Gracefully handle if settings.attributes.connections is undefined.
 		settings.attributes = {
 			...settings.attributes,
-			__experimentalConnections: {
+			connections: {
```
Comment (michalczaplinski) [inline packages/block-editor/src/hooks/custom-fields.js]: I'm not sure why these changes show up in GithHub's UI. They are already in the other PR: https://github.com/WordPress/gutenberg/pull/53241

File: lib/experimental/blocks.php (hunk 6 above)
Comment (SantosGuillamot) [reply lib/experimental/blocks.php]: Yep, I was thinking about names using this sentence: **we are connecting block attributes to values that might change depending on the context (post_id, parent block…)**. But not sure yet which is the best abstraction 😄

File: packages/block-editor/src/hooks/custom-fields.js (hunk 7 above)
Comment (SantosGuillamot) [reply packages/block-editor/src/hooks/custom-fields.js]: Maybe we need to rebase/merge the other PR branch. If I am not mistaken the last merge was done before the change you mention.

File: lib/experimental/blocks.php (hunk 6 above)
Comment (michalczaplinski) [reply lib/experimental/blocks.php]: Hmm, I would just go with "block connections" as a name for the whole thing (like I use it in https://github.com/WordPress/gutenberg/pull/53247/files#diff-be8a354ce9c01e85c0fdf04c75d76880cdc6c81992db7a86a7a0515dc11a9833R135) and "connection sources" for the thing that does the connecting to meta fields/parent blocks/other DBs, etc.

We can always change it in the future.

If you don't object, then I ll go ahead and change it.

Comment (SantosGuillamot) [reply lib/experimental/blocks.php]: Sure! We can always change the naming later if we agree on a different one.

**Review by SantosGuillamot — APPROVED** [review]
Looks good to me for this first iteration 🙂

**Comment by michalczaplinski** [comment]
Great, I've just updated the naming to consistently use:
- "connection sources" as the thing doing the connecting to meta fields/other sources in the future
- "block connections" as the name for the whole feature.

I'll first merge #53241 and then this one.

File: packages/block-editor/src/hooks/custom-fields.js (hunk 7 above)
Comment (michalczaplinski) [reply packages/block-editor/src/hooks/custom-fields.js]: yep, I had to merge `experiment/add-custom-connectinons-inspector-control` into this one before pushing again 🙂

File: lib/experimental/blocks.php [hunk 8]
```diff
@@ -121,3 +121,107 @@ function gutenberg_register_metadata_attribute( $args ) {
...
+		$block_type         = $block_instance->block_type;
+
+		// Allowlist of blocks that support block connections.
+		// Currently, we only allow the following blocks and attributes:
+		// - Paragraph: content.
+		// - Image: url.
+		$blocks_attributes_allowlist = array(
+			'core/paragraph' => array( 'content' ),
+			'core/image'     => array( 'url' ),
+		);
```
Comment (youknowriad) [inline lib/experimental/blocks.php:146]: Can we retrieve this from a block support instead of a hardcoded list.

Comment (youknowriad) [reply lib/experimental/blocks.php:146]: Actually it's the user that should define which attribute to connect, so this is coming from the block attribute value  normally no?

File: lib/experimental/blocks.php [hunk 9]
```diff
@@ -121,3 +121,107 @@ function gutenberg_register_metadata_attribute( $args ) {
...
+				continue;
+			}
+
+			// If the attribute does not specify the name of the custom field, skip it.
+			if ( ! isset( $attribute_value['value'] ) ) {
+				continue;
+			}
+
+			// Get the content from the connection source.
+			$custom_value = $connection_sources[ $attribute_value['source'] ](
```
Comment (youknowriad) [inline lib/experimental/blocks.php:194]: This is weird, I'd have preferred a unique name for all connection, like `get_value` or `apply` or something.

File: packages/block-library/src/paragraph/block.json [hunk 10]
```diff
@@ -7,6 +7,7 @@
 	"description": "Start with the basic building block of all narrative.",
 	"keywords": [ "text" ],
 	"textdomain": "default",
+	"usesContext": [ "postId" ],
```
Comment (youknowriad) [inline packages/block-library/src/paragraph/block.json:10]: I wonder whose responsibility is it to do that. In other words, the block is not using the "postId" explicitly, so why add the "useContext" here. The "connection framework" should inject/filter this somehow.

File: lib/experimental/blocks.php (hunk 8 above)
Comment (michalczaplinski) [reply lib/experimental/blocks.php:146]: See my comment in https://github.com/WordPress/gutenberg/pull/53241#discussion_r1293545604 :) 

Yes, but for now we want to limit this to core blocks and those two attributes. We'll expand it in the future. 

So far it's not well-defined how users are going to opt-in to connecting their custom blocks' attributes to custom fields.

File: lib/experimental/blocks.php (hunk 9 above)
Comment (michalczaplinski) [reply lib/experimental/blocks.php:194]: Oh, maybe the confusion is coming from this line: https://github.com/WordPress/gutenberg/blob/676f25961bfab695bacb5dd7ebb10443e49bce23/lib/experimental/connection-sources/index.php#L9

I should have removed the `"name"` because it serves no purpose at the moment.

So in your PR, there was an assumption that there's one file per connection source (`attributes-custom-sources/meta.php`, `attributes-custom-sources/some-other-source.php`) and here they are all in under keys of one array. I say "all" but there is only one type of connection source, anyway 😅 .

Comment (youknowriad) [reply lib/experimental/blocks.php:194]: ah ok, I guess depending on the complexity we could split but agree, seems like implementation detail for now but let's aim for the simplest. I do expect us to allow registering sources at some point so I think "one config array per source" is probably better ultimately (as opposed to all configs in a single array).

File: packages/block-library/src/paragraph/block.json (hunk 10 above)
Comment (michalczaplinski) [reply packages/block-library/src/paragraph/block.json:10]: Do you mean that in the context of Core blocks or custom blocks or both?

Comment (youknowriad) [reply packages/block-library/src/paragraph/block.json:10]: both

Comment (michalczaplinski) [reply packages/block-library/src/paragraph/block.json:10]: OK! For now, I think that for core blocks this is fine. But you're right that for custom blocks we'll have to come up with a proper mechanism here. Are you all right with that?

Comment (youknowriad) [reply packages/block-library/src/paragraph/block.json:10]: Yeah I think it's part of the definition of a "custom source" to say which change it needs to make to block registration. I think we should use it for both core and third-party blocks but yeah it's fine to start without it.

Comment (gziolo) [reply packages/block-library/src/paragraph/block.json:10]: For now, you could update `usesContext` and pass `postId` in PHP when processing `gutenberg_render_block_connections`. This way, it won't get unnecessarily wired for all Paragraph block instances through `block.json` definition.

Comment (SantosGuillamot) [reply packages/block-library/src/paragraph/block.json:10]: Even though this is the first implementation and custom blocks are not supported yet, I believe it'd be better to start creating mechanisms that would work for both if possible. Mainly because we already know we will want to support that in the future.

Comment (talldan) [reply packages/block-library/src/paragraph/block.json:10]: Coming in a bit late here.

The current work on partial syncing (#54233) is using the store actions/selectors for attributes to inject/update values, but it might be difficult to pull this off if block context is being used for custom fields, as I believe it uses react context in JS.

While they're still proofs of concepts, the current work for custom fields/partial syncing does need to remain compatible, so it might be good to explore ways to read/write custom field values in the editor soon though so that we have a clear idea of the problems that need to be solved.

File: lib/experimental/blocks.php [hunk 11]
```diff
@@ -121,3 +121,107 @@ function gutenberg_register_metadata_attribute( $args ) {
...
+					// TODO: In the future, when blocks other than Paragraph and Image are
+					// supported, we should build the full query from CSS selector.
+					'tag_name' => $block_type->attributes[ $attribute_name ]['selector'],
+				)
+			);
+			if ( ! $found ) {
+				return $block_content;
+			};
+			$tag_name     = $tags->get_tag();
+			$markup       = "<$tag_name>$custom_value</$tag_name>";
```
Comment (mcsf) [inline lib/experimental/blocks.php:211]: This line is scary. Any plans for escaping? It's not trivial, since the goal is to be able to inject rich text and not just simple values.

Comment (dmsnell) [reply lib/experimental/blocks.php:211]: thanks for catching this @mcsf. this definitely should be `"<$tag_name>" . esc_html( $custom_value ) . "</$tag_name>"` for now.

actually @michalczaplinski let's talk about how to sub-class the Tag Processor so this is cleaner all over

Comment (SantosGuillamot) [reply lib/experimental/blocks.php:211]: Thanks for raising it! We should definitely do that. Keep in mind that this code was considered experimental and we'll need to refactor it.

Comment (mcsf) [reply lib/experimental/blocks.php:211]: Definitely, but all too often lines of experimental code somehow make their way into production code. :) I'd really patch ASAP, even if just with `esc_html`.

Comment (michalczaplinski) [reply lib/experimental/blocks.php:211]: Thanks for catching this! 

I want to take a holistic look at all the code shipped behind the experimental flag. It's very likely most of it will be replaced altogether. The team is on support rotation this week so it'll have to wait till next week at least 🙂 

@dmsnell yes, let's talk!
